X Axis labels for bar chart

sepans opened the pull request:
Fixes [GALL-1645](https://artsyproduct.atlassian.net/browse/GALL-1645)

<img width="873" alt="Screen Shot 2019-04-24 at 2 50 12 PM" src="https://user-images.githubusercontent.com/687513/56685593-4df1bf00-66a0-11e9-876c-f1fb97e05d3c.png">

damassi commented on packages/palette/src/elements/BarChart/Bar.tsx:
```diff
@@ -137,6 +138,7 @@ const HighlightLabel = ({
 export const Bar = ({
   heightPercent,
   label,
+  xAxisLabel,
```
Maybe `axisLabelX` / `AxisLabelX`?

artsyit commented:
Deploy preview for *artsy-palette* ready!

Built with commit 8aad17a23986ad2e49c28c3fc9f7b8f3f6645e1b

https://deploy-preview-392--artsy-palette.netlify.com

sepans commented on packages/palette/src/elements/BarChart/Bar.tsx:
```diff
@@ -185,6 +188,16 @@ export const Bar = ({
+
+const XAxisLabel = styled(Sans)`
+  position: absolute;
+  bottom: -${space(2)}px;
```
This part is a bit too much a hack and bothers me. It is because the label is inside the `<BarBox>` to have the horizontal alignment right but then had to do a negative margin bottom here (which is dependent on the height of the label).

ds300 replied:
you could do `top: calc(100% + ${marginBetweenLabelAndBaseLine});`

ds300 commented on packages/palette/src/elements/BarChart/BarChart.tsx:
```diff
@@ -115,13 +116,17 @@ export const BarChart = ({ bars, minLabel, maxLabel }: BarChartProps) => {
                 <Bar
                   key={index}
                   heightPercent={heightPercent}
                   label={coerceLabel(label)}
```
in hindsight this probably should have been called tooltip 😞

ds300 commented on packages/palette/src/elements/BarChart/Bar.tsx:
```diff
@@ -185,6 +188,16 @@ export const Bar = ({
+  position: absolute;
+  top: calc(100% + ${space(0.5)}px);
+  width: 100%;
+  text-align: center;
```
I think to be more flexible it would make sense to center this whole box instead of just the text. That would mean the box is not constrained to be the same width as the bar, and then the text will still be centered even if the text is wider than the bar.

That is, unless you want to use word wrapping in these labels? Then you'd need an explicit width like how it's done now.

sepans replied:
Yes good catch looking at the designs we don't want the labels to be constrained by the bar width. cc: @owendodd and @samchieng:

<img width="647" alt="Screen Shot 2019-04-25 at 10 31 03 AM" src="https://user-images.githubusercontent.com/687513/56743865-64058b00-6745-11e9-8982-7db74818d1bd.png">

sepans replied:
pfffff, I thought it was simple but haven't figured it out yet. Is it possible to center align the label without removing `position: absolute;`? If not it requires a good amount of change to the way bars are positioned I guess.

ds300 replied:
try
```
left: 50%;
transform: translateX(-50%);
```

sepans replied:
I got that with https://github.com/artsy/palette/pull/392/commits/0ff165567c44cc0a9f0c34e4561e96b3d9e34890#diff-2e24d9a0ed8b8746d5fa5d1333da123dR231:

<img width="825" alt="Screen Shot 2019-04-25 at 12 05 16 PM" src="https://user-images.githubusercontent.com/687513/56751775-d5006f00-6754-11e9-813a-d22cb71237ef.png">

The only tricky thing was that I had to add https://github.com/artsy/palette/pull/392/commits/0ff165567c44cc0a9f0c34e4561e96b3d9e34890#diff-710f36809acea9f66cf6ae101cf56b66R13 to remove the extra padding on the right side of the chart created but `width: 200%;`

Maybe I was missing something but by some reason 👆 didn't quite work:

<img width="828" alt="Screen Shot 2019-04-25 at 12 23 00 PM" src="https://user-images.githubusercontent.com/687513/56751823-ec3f5c80-6754-11e9-9791-c4825467c3e4.png">

ds300 replied:
Hmm I tried the left + transform trick and it looked fine.

Also while playing around I noticed that these labels aren't visible for me because of overflow issues. They don't add to the height of the bar chart, but they probably should.

It might be best to rethink how we're doing this. I'm imagining a parallel row of equal-sized flex containers underneath the row of bars which center their children using flexbox and align items using `baseline`, does that make sense to you?

```
Flex row ----->

+-------------+
|         +-+ |
| +-+     | | |
| | |     | | |
| | | +-+ | | |
| | | | | | | |
| | | | | | | |
| | | | | | | |
| +-+ +-+ +-+ |
+-------------+
+-------------+
| +-+ +-+ +-+ |
| +-+ +-+ +-+ |
+-------------+

Flex row ----->

```

ds300 replied:
actually just tried that out and don't think it'd work because flexbox doesn't center overflowing content 😞

ds300 commented:
@sepans Another issue I see with this is that the x-axis labels are mutually exclusive with the existing min+max label props. It would be cool to capture that in the api design. Could do it by deprecating `minLabel` and `maxLabel` and creating a new polymorphic `xAxisLabels` prop that can capture both situations and any future requirements.

e.g
![image](https://user-images.githubusercontent.com/1242537/57072531-fa9df300-6cd5-11e9-8f86-4f7336d03f06.png)

sepans commented:
@ds300 👆 is a great idea and I went ahead and changed it until  I realized in order to have x-axis labels for some bars an array like `[undefined, '200', undefined, '100'] needs to be passed to the `xAxisLabels`.

Also thinking it over the mutual exclusiveness of `min/maxLabel` and `xAxisLabels` is just a visual constraint and not semantic (e.g. hypothetically if in later design iterations the position of `min/maxLabel` changes they are no longer mutually exclusive) so I won't be super worried if it is not enforced in the component API.

ds300 commented:
Ah yeah that's a great point.

Did you have time to look into the vertical height issue?

sepans commented:
I tried a couple of different things but haven't found a solution for that yet. 😥

sepans commented:
@ds300 I tried another hybrid approach which is not perfect but better. I explained it in the commit message and here is the screenshot:

<img width="845" alt="Screen Shot 2019-05-15 at 11 50 39 AM" src="https://user-images.githubusercontent.com/687513/57790077-0e9b1900-7708-11e9-9fd1-dfe2e4426e55.png">

let me know what you think 🙏

ds300 commented:
Thanks for your patience on this :pray: That looks like a good compromise to get stuff done and move on for now.

I noticed some unintended scroll behaviour.

![nested-scroll](https://user-images.githubusercontent.com/1242537/57845648-4ff3fd00-77ca-11e9-8da6-2e03a2a0b6ea.gif)

I looked into this for a few minutes but couldn't figure out what the issue was. Do you mind having a deeper look?

sepans commented:
Thanks @ds300 for catching that. It was actually caused by the large font size of the `December 30` and `April` (that `min-height` hack only takes care of the labels up to a certain line height). Since large labels are not in out design system I just decreased the font size to fix it. 

And also noticed an extra `overflow: hidden` tag that that I had added and was not really needed and removed it.

artsyit commented:
<!-- GITHUB_RELEASE COMMENT: released -->
:rocket: PR was released in v4.12.0 :rocket:
